File: encoders/api/python/eqep.py
# API for the TI eQEP hardware driver I wrote

# We need OS operations for this
import os, select

class eQEP(object):
    # Modes
    MODE_ABSOLUTE = 0
    MODE_RELATIVE = 1
    
    # eQEP Controller Locations
    eQEP0 = "/sys/devices/ocp.2/48300000.epwmss/48300180.eqep"
    eQEP1 = "/sys/devices/ocp.2/48302000.epwmss/48302180.eqep"
    eQEP2 = "/sys/devices/ocp.2/48304000.epwmss/48304180.eqep"

    # Set the mode of the eQEP hardware
    def set_mode(self, mode):
        # Open the mode attribute file
        attribute = open(self.path + "/mode", "w")
        
        # Write the desired mode into the file
        attribute.write(str(mode))
        
        # Close the file
        attribute.close()
        
    # Get the mode of the eQEP hardware
    def get_mode(self):
        # Open the attribute file
        attribute = open(self.path + "/mode", "r")
        
        # Get the value
        mode = int(attribute.readline())
        
        # Close the attribute
        attribute.close()
        
        # Return the mode
        return mode

    # Set the unit timer period of the eQEP hardware
    def set_period(self, period):
        # Open the mode attribute file
        attribute = open(self.path + "/period", "w")
        
        # Write the desired mode into the file
        attribute.write(str(period))
        
        # Close the file
        attribute.close()
        
    # Get the unit timer period of the eQEP hardware
    def get_period(self):
        # Open the attribute file
        attribute = open(self.path + "/period", "r")
        
        # Get the value
        period = int(attribute.readline())
        
        # Close the attribute
        attribute.close()
        
        # Return the mode
        return period
        
    # Set the current position of the encoder hardware
    def set_position(self, position):
        # Open the mode attribute file
        attribute = open(self.path + "/position", "w")
        
        # Write the desired mode into the file
        attribute.write(str(position))
        
        # Close the file
        attribute.close()
        
    # Get the immediate position of the encoder hardare
    def get_position(self):
        # Open the attribute file
        attribute = open(self.path + "/position", "r")
        
        # Get the value
        position = int(attribute.readline())
        
        # Close the attribute
        attribute.close()
        
        # Return the mode
        return position
        
    # Poll the position, returns when new data is available
    def poll_position(self):
        # Poll the position file
        self.poller.poll(-1)
        
        # Seek to the beginning of the file to get the data
        os.lseek(self.fd, 0, 0)
        
        # Return the position
        return int(os.read(self.fd, 16))
    
    # Constructor - specify the path and the mode
    def __init__(self, path, mode):
        # Base path of the eQEP sysfs entry (ex. /sys/devices/ocp.2/48302000.epwmss/48302180.eqep)
        self.path = path;
        
        # Set the mode
        self.set_mode(mode)

        # Reset the position
        self.set_position(0)
        
        # Setup polling system
        self.fd = os.open(self.path + "/position", os.O_RDONLY, os.O_NONBLOCK)
       
        # Create the poll object
        self.poller = select.poll()
        self.poller.register(self.fd, select.POLLPRI)  
        
    # Deconstructor
    def __del__(self):
        # Cleanup polling system
        self.poller.unregister(self.fd)
        os.close(self.fd)   
        
        

File: encoders/api/c++/eqep.cpp
/*
 * TI eQEP driver interface API
 *
 * Copyright (C) 2013 Nathaniel R. Lewis - http://nathanielrlewis.com/
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.
 *
 */
 
// Pull in our eQEP driver definitions
#include "eqep.h"
 
// Language dependencies
#include <cstdint>
#include <cstdlib>
#include <cstdio>

// POSIX dependencies
#include <unistd.h>
#include <fcntl.h>
#include <poll.h> 
#include <sys/types.h> 
#include <sys/stat.h> 
 
// Constructor for eQEP driver interface object
eQEP::eQEP(std::string _path, eQEP::eQEP_Mode _mode)
    : path(_path)
{
    // Set the mode of the hardware
    this->set_mode(_mode);
    
    // Reset the position
    this->set_position(0);
}

// Set the position of the eQEP hardware
void eQEP::set_position(int32_t position)
{
    // Open the file representing the position
    FILE *fp = fopen((this->path + "/position").c_str(), "w");
    
    // Check that we opened the file correctly
    if(fp == NULL)
    {
        // Error, break out
        std::cerr << "[eQEP " << this->path << "] Unable to open position for write" << std::endl;
        return;
    }

    // Write the desired value to the file
    fprintf(fp, "%d\n", position);
    
    // Commit changes
    fclose(fp);
}

// Set the period of the eQEP hardware
void eQEP::set_period(uint64_t period)
{
    // Open the file representing the position
    FILE *fp = fopen((this->path + "/period").c_str(), "w");
    
    // Check that we opened the file correctly
    if(fp == NULL)
    {
        // Error, break out
        std::cerr << "[eQEP " << this->path << "] Unable to open period for write" << std::endl;
        return;
    }

    // Write the desired value to the file
    fprintf(fp, "%llu\n", period);
    
    // Commit changes
    fclose(fp);
}

// Set the mode of the eQEP hardware
void eQEP::set_mode(eQEP::eQEP_Mode _mode)
{
    // Open the file representing the position
    FILE *fp = fopen((this->path + "/mode").c_str(), "w");
    
    // Check that we opened the file correctly
    if(fp == NULL)
    {
        // Error, break out
        std::cerr << "[eQEP " << this->path << "] Unable to open mode for write" << std::endl;
        return;
    }

    // Write the desired value to the file
    fprintf(fp, "%u\n", _mode);
    
    // Commit changes
    fclose(fp);
}

// Get the position of the hardware
int32_t eQEP::get_position(bool _poll)
{
    // Position temporary variable
    int32_t       position;
    char          dummy;
    struct pollfd ufd;
    
    // Do we want to poll? 
    if(_poll)
    {
        // Open a connection to the attribute file.
        if ((ufd.fd = open((this->path + "/position").c_str(), O_RDWR)) < 0)
        {
            // Error, break out
            std::cerr << "[eQEP " << this->path << "] Unabled to open position for polling" << std::endl;
            return 0 ;
        }

        // Dummy read
        read(ufd.fd, &dummy, 1);
        
        // Poll the port
        ufd.events = POLLPRI;
        if (poll(&ufd, 1, -1) < 0)
        {
            // Error, break out
            std::cerr << "[eQEP " << this->path << "] Error occured whilst polling" << std::endl;
            close(ufd.fd);
            return 0 ;
        }
    } 
    
    // Read the position
    FILE *fp = fopen((this->path + "/position").c_str(), "r");
    
    // Check that we opened the file correctly
    if(fp == NULL)
    {
        // Error, break out
        std::cerr << "[eQEP " << this->path << "] Unable to open position for read" << std::endl;
        close(ufd.fd);
        return 0 ;
    }

    // Write the desired value to the file
    fscanf(fp, "%d", &position);
    
    // Commit changes
    fclose(fp);
    
    // If we were polling, close the polling file
    if(_poll)
    {
        close(ufd.fd);
    }
    
    // Return the position
    return position;
}

// Get the period of the eQEP hardware
uint64_t eQEP::get_period()
{
    // Open the file representing the position
    FILE *fp = fopen((this->path + "/period").c_str(), "r");
    
    // Check that we opened the file correctly
    if(fp == NULL)
    {
        // Error, break out
        std::cerr << "[eQEP " << this->path << "] Unable to open period for read" << std::endl;
        return 0 ;
    }

    // Write the desired value to the file
    uint64_t period = 0;
    fscanf(fp, "%llu", &period);
    
    // Commit changes
    fclose(fp);
    
    // Return the period
    return period;
}

// Get the mode of the eQEP hardware
eQEP::eQEP_Mode eQEP::get_mode()
{
    // Open the file representing the position
    FILE *fp = fopen((this->path + "/mode").c_str(), "r");
    
    // Check that we opened the file correctly
    if(fp == NULL)
    {
        // Error, break out
        std::cerr << "[eQEP " << this->path << "] Unable to open mode for read" << std::endl;
        return eQEP::eQEP_Mode_Error;
    }

    // Write the desired value to the file
    eQEP::eQEP_Mode mode;
    fscanf(fp, "%u", &mode);
    
    // Commit changes
    fclose(fp);
    
    // Return the mode
    return mode;
}
